### PR TITLE
Daniel record 2

### PR DESCRIPTION
**Pull request by dkrasner**
### Main Points ###

This PR represents a complete rewrite of how we interface with instructions and the callstack. There is no longer an 'intermediary' mapper worksheet that essentially worked as the callstack editor, but a [CommandInterface](https://github.com/dkrasner/worksheet/compare/daniel-record-2?expand=1#diff-7026da4d0a99e5a7706d346a91ba4231bbd8e76838e8a903003a5c1b5243fd69) custom element that allows for command selection, argument editing, documentation, trying things out, saving to the callstack etc:

image

With this PR the source(s) and target worksheet are linked directly, creating a `ws-connection` element with the corresponding sources and targets. When a link happens the target worksheet acquires callstack related interfaces (see more on this below):
image

There are some new buttons: the record button (which toggles red/green) puts the target in record mode (it should really put the connection in record mode, see comments below), and here drag&drop selection trigger the opening of the command interface. (It was pointed out to me that we shouldn't call it "record" since that makes people think of screen-recordings and the like, whereas I was thinking of VIM-record which no one knows about...) 

In view of all this the worksheet no longer interacts with the callstack directly, and no longer deals with hacky pre-processing of the instructions. All of this code has been removed. There is still callstack related functionality, like stepping, running and viewing the stack but these are [passed to the connection element](https://github.com/dkrasner/worksheet/compare/daniel-record-2?expand=1#diff-79bbbea1112592ced9050ee2d67f8aa74f71d335fa20acd1d59bfa646f569271R118). Whether the connection element or the command interface or something else should be dealing with the callstack is to be sorted out. I think this will largely depend on whether we introduce a visual element for the connection and what it looks like and what it does. 

Other things:
We've lost the step through highlighting, although I left the [method](https://github.com/dkrasner/worksheet/blob/e2a797b7d1f43c071dbb99dccb1710992b93d1ae/js/Worksheet.js#L658) for future reference. I think we'll need to bring something like this back. 
Given how structured the command interface is it gives you the instruction directy. However, I have not changed how the interpreter works and so at the moment this instruction is [combined into a string](https://github.com/dkrasner/worksheet/compare/daniel-record-2?expand=1#diff-7026da4d0a99e5a7706d346a91ba4231bbd8e76838e8a903003a5c1b5243fd69R333) only later to the parsed into the same instruction. First, I didn't want to change our interpreter for this PR and didn't want to think about how to deal with editing the callstack (as discussed in issue #34)

I've opened a number of issues to think about whether this should go.

### CODE REVIEW

File: js/CommandInterface.js [hunk 1]
```diff
@@ -0,0 +1,347 @@
...
+        doit_button.removeEventListener("click", this.onDoIt);
+        const saveit_button = this.shadowRoot.querySelector("button#save-it");
+        saveit_button.removeEventListener("click", this.onSaveIt);
+    }
+
+
+    onComamndSelection(event){
+        const name = event.target.value;
+        const editor = this.shadowRoot.querySelector("#editor");
+        const command_info = this.interpreter.command_registry[name];
```
Comment (darth-cheney) [inline js/CommandInterface.js:207]: There seems to be a mix in here of snake_case and camelCase. Should we update to stick with camelCase? (maybe you were switching between this and a python project)

File: js/WSConnection.js [hunk 2]
```diff
@@ -4,23 +4,38 @@
...
         this.onWorksheetMoved = this.onWorksheetMoved.bind(this);
+        this.openCommandInterface = this.openCommandInterface.bind(this);
+        this.step = this.step.bind(this);
+        this.run = this.run.bind(this);
+        this.inspectCallstack = this.inspectCallstack.bind(this);
     }
 
     connectedCallback() {
         if (this.isConnected) {
+            this.setAttribute("id", window.crypto.randomUUID());
```
Comment (darth-cheney) [inline js/WSConnection.js:33]: This should be fine in our use case, but a new uuid will be generated any time this element gets moved around (I know that's unlikely to happen at this point). The alternative would be to set an instance property directly in the constructor and then set the attribute to that value on any connectedCallback

**Review by darth-cheney — CHANGES_REQUESTED** [review]
Aside from some inline comments I made, the code itself looks good.
  
There are a couple of things to note. 
  
First, when I load the `commandInterface.html` example, I have the following error in the console:
Screen Shot 2022-09-14 at 5 21 46 PM
  
I'm assuming this has something to do with the element being connected but an interpreter instance is not yet initialized at that time.
  
The second is that it's not clear to me how to edit connections or get the commandInterfacce to appear in the `mapper.html` example. I tried various kinds of clicks, etc, but perhaps that example could use some instructions on the background of the page. Or maybe you can just add some instructions on how to test this in the comments of the PR.

**Comment by dkrasner** [comment]
> Aside from some inline comments I made, the code itself looks good.
> 
> There are a couple of things to note.
> 
> First, when I load the `commandInterface.html` example, I have the following error in the console: Screen Shot 2022-09-14 at 5 21 46 PM
> 
> I'm assuming this has something to do with the element being connected but an interpreter instance is not yet initialized at that time.
> 
> The second is that it's not clear to me how to edit connections or get the commandInterfacce to appear in the `mapper.html` example. I tried various kinds of clicks, etc, but perhaps that example could use some instructions on the background of the page. Or maybe you can just add some instructions on how to test this in the comments of the PR.

oh yeah those .html examples are all old, even in the interface one which was just used to have something to look at for a day, and probably should be removed - which i can do now

File: js/WSConnection.js (hunk 2 above)
Comment (dkrasner) [reply js/WSConnection.js:33]: this just temp to deal with callstack viewing, i am not sure how we should really identify a connection in the future whether really the callstack should carry the id etc

File: js/CommandInterface.js (hunk 1 above)
Comment (dkrasner) [reply js/CommandInterface.js]: fixed [here](https://github.com/dkrasner/worksheet/pull/35/commits/a93651474ccb3e1d8328d8288e38eedd044df09a)

**Comment by dkrasner** [comment]
@darth-cheney i updated mapper.html to reflect the current and remove interface html

**Pull request merged**